### PR TITLE
Debounced Autosave Individual Grade

### PR DESCRIPTION
**Pull request by bivanalhar**
Issue: currently, when grader is in the midst of grading the scores, the grade won't be saved until they clicked on the button "Save Grade" at the bottom of the page. This might create some nuisance and unexpected behaviour when grader accidentally close the page and then upon revisiting, found out the grades are not saved yet.

What's fixed within this PR:
1. Different behavior of grade saving when the workflow state of the submission is Graded/Published as compared to the rest. 
- When the workflow state is Graded/Published, then the autosave won't be activated and hence if user wished to modify the grade, he/she needs to go all the way to the bottom and click Save Grade. 
- Otherwise, autosave will take place and the grade will be automatically saved upon changing

2. Behavior of autosaving the grade
- basically, what user types the last will be what's being saved to the system. If user didn't input anything (cancelling the grade input he/she has inputted previously), then the grade will be reverted back to null. Changing into any grade, for example 3, also will automatically invoke the autosave into 3.

3. Fix displaying error message related with answer
- Previously, there's a bug in which if the grade submitted is null after the submission is either graded or published, even though request is rejected by the backend, the error message won't be appearing properly in the frontend

### CODE REVIEW

File: client/app/bundles/course/assessment/submission/actions/index.js [hunk 1]
```diff
@@ -484,15 +485,70 @@ export function saveGrade(submissionId, grades, exp, published) {
...
   return (dispatch) => {
-    dispatch({ type: actionTypes.SAVE_GRADE_REQUEST });
+    dispatch({ type: actionTypes.SAVE_ALL_GRADE_REQUEST });
 
     return CourseAPI.assessment.submissions
       .update(submissionId, payload)
       .then((response) => response.data)
       .then((data) => {
-        dispatch({ type: actionTypes.SAVE_GRADE_SUCCESS, payload: data });
+        dispatch({ type: actionTypes.SAVE_ALL_GRADE_SUCCESS, payload: data });
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/actions/index.js:494]: You are renaming `SAVE_GRADE_REQUEST`, `SAVE_GRADE_SUCCESS` and `SAVE_GRADE_FAILURE`. Have you checked other places that uses the same action names?

File: client/app/bundles/course/assessment/submission/constants.js [hunk 2]
```diff
@@ -161,8 +161,11 @@ const actionTypes = mirrorCreator([
   'RESET_REQUEST',
   'RESET_SUCCESS',
   'RESET_FAILURE',
+  'SAVE_ALL_GRADE_REQUEST',
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/constants.js:164]: nit: group them together

File: client/app/bundles/course/assessment/submission/actions/index.js [hunk 3]
```diff
@@ -474,7 +475,7 @@ export function importFiles(answerId, answerFields, language, setValue) {
   };
 }
 
-export function saveGrade(submissionId, grades, exp, published) {
+export function saveAllGrade(submissionId, grades, exp, published) {
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/actions/index.js:478]: nit: `saveAllGrades` - update for everything related. Use plural.

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEmptyForm.jsx [hunk 4]
```diff
@@ -74,7 +74,7 @@ const SubmissionEmptyForm = (props) => {
       <Button
         color="primary"
         disabled={isSaving}
-        onClick={handleSaveGrade}
+        onClick={() => handleSaveAllGrade()}
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEmptyForm.jsx:77]: You can pass `handleSaveAllGrades` directly since it is already a function without any argument. No need to recreate another function for this. (Explain why)

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx [hunk 5]
```diff
@@ -404,7 +411,7 @@ const SubmissionEditStepForm = (props) => {
       <Button
         color="primary"
         disabled={isSaving}
-        onClick={handleSaveGrade}
+        onClick={() => handleSaveAllGrade()}
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditStepForm.jsx:414]: Same comment as the previous one.

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx [hunk 6]
```diff
@@ -492,7 +499,7 @@ const SubmissionEditForm = (props) => {
       <Button
         color="primary"
         disabled={isSaving}
-        onClick={handleSaveGrade}
+        onClick={() => handleSaveAllGrade()}
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx:502]: Same comment

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx [hunk 7]
```diff
@@ -401,7 +402,13 @@ const SubmissionEditForm = (props) => {
     const editable = !attempting && graderView;
     const visible = editable || published;
 
-    return visible ? <QuestionGrade editable={editable} id={id} /> : null;
+    return visible ? (
+      <QuestionGrade
+        editable={editable}
+        handleSaveGrade={() => handleSaveGrade(id)}
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx:408]: Since `id` is also passed in to this component, it is better to just pass the `handleSaveGrade` directly and when the function is invoked inside `QuestionGrade`, you can pass the `id`. Also, this reduces the need to recreate a new function.

File: client/app/bundles/course/assessment/submission/containers/QuestionGrade.jsx [hunk 8]
```diff
@@ -77,14 +77,16 @@ class VisibleQuestionGrade extends Component {
             className="w-40"
             hiddenLabel
             inputProps={{ className: 'grade' }}
-            onBlur={(e) => this.processValue(e.target.value)}
+            onBlur={(e) => {
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/containers/QuestionGrade.jsx:80]: You might not be able to use `useDebounce` here since it is a class component. If you want, you can convert it to functional component. I think it is a good practice. Do convert this, make this component tidier and use typescript.

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx [hunk 9]
```diff
@@ -352,7 +375,8 @@ class VisibleSubmissionEditIndex extends Component {
         handleAutogradeSubmission={() => this.handleAutogradeSubmission()}
         handleMark={() => this.handleMark()}
         handlePublish={() => this.handlePublish()}
-        handleSaveGrade={() => this.handleSaveGrade()}
+        handleSaveAllGrade={() => this.handleSaveAllGrade()}
+        handleSaveGrade={(id) => this.handleSaveGrade(id)}
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx:379]: You can just pass `this.handleSaveGrade`?

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx [hunk 10]
```diff
@@ -122,7 +123,27 @@ class VisibleSubmissionEditIndex extends Component {
...
+      exp,
+      questions,
+    } = this.props;
+    dispatch(
+      saveGrade(
+        params.submissionId,
+        [grading[id]],
+        Object.values(grading),
+        exp,
+        false,
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx:144]: How do you know this is always `false`?

File: client/app/bundles/course/assessment/submission/actions/index.js [hunk 11]
```diff
@@ -484,15 +485,70 @@ export function saveGrade(submissionId, grades, exp, published) {
...
+            },
+          })),
+        };
+
+        dispatch({
+          type: actionTypes.SAVE_GRADE_SUCCESS,
+          payload: dataWithPrefilledGrades,
+        });
+        dispatch(
+          setNotification(translations.updateIndividualSuccess, questionTitle),
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/actions/index.js:549]: Instead of using notification bar, consider adding autosaving icon/status like what you proposed for answer autosaving?

File: client/app/bundles/course/assessment/submission/actions/index.js [hunk 12]
```diff
@@ -484,15 +485,70 @@ export function saveGrade(submissionId, grades, exp, published) {
...
+        dispatch(
+          setNotification(translations.updateFailure, buildErrorMessage(error)),
+        );
+      });
+  };
+}
+
+export function saveGrade(
+  submissionId,
+  grades,
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/actions/index.js:508]: Instead of passing an array that only contains 1 element, suggest passing the raw data and build the payload inside. That way, someone who invokes this function does not need to know the specific format to be passed to the payload.

File: client/app/bundles/course/assessment/submission/actions/index.js [hunk 13]
```diff
@@ -484,15 +485,70 @@ export function saveGrade(submissionId, grades, exp, published) {
...
+  const mapAnswerIdToAnswer = arrayToObjectById(allGrades);
+
+  return (dispatch) => {
+    dispatch({ type: actionTypes.SAVE_GRADE_REQUEST });
+
+    return CourseAPI.assessment.submissions
+      .update(submissionId, payload)
+      .then((response) => response.data)
+      .then((data) => {
+        const dataWithPrefilledGrades = {
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/actions/index.js:531]: use `produce` from `immer`.

Comment (ekowidianto) [reply client/app/bundles/course/assessment/submission/actions/index.js:531]: Also, update only the grade that is saved to the backend. This function took the grade of 1 question. To avoid weird unexpected behaviour, it should only update the grade of that particular question in redux.

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Some comments

File: client/app/bundles/course/assessment/submission/actions/index.js (hunk 1 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/actions/index.js:494]: yes, thus far, in the other places (other than actually saving the grade, for example in annotation, explanation, etc.), I assume both SAVE_ALL_GRADE_SUCCESS and SAVE_GRADE_SUCCESS have the same effect on the states.

Though, this warrants the further inspection (thus far I didn't see any difference in terms of effect) on whether they are actually supposed to have the same effect. I will check on it further later.

Comment (ekowidianto) [reply client/app/bundles/course/assessment/submission/actions/index.js:494]: What I mean is that, previously, saveGrade uses `SAVE_GRADE_SUCCESS` . This action also exists in multiple reducers (ie annotations, answers, attachment, etc). Once you rename to `SAVE_ALL_GRADE_SUCCESS`, why `SAVE_ALL_GRADE_SUCCESS` is not added to the other reducers as well?

Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/actions/index.js:494]: oh I see. yes regarding that one, it's the mistake from my part that I didn't add `SAVE_ALL_GRADE_SUCCESS` to all the places (mainly because I was focused on saving grades, unaware that the other reducers is using this one). In the current commit I have just pushed (it's unfinished), I have added that to other reducers as well..

but it might need some further inspection, which I will do later

File: client/app/bundles/course/assessment/submission/translations.ts [hunk 14]
```diff
@@ -262,6 +262,10 @@ const translations = defineMessages({
     id: 'course.assessment.submission.updateSuccess',
     defaultMessage: 'Submission updated successfully.',
   },
+  updateIndividualSuccess: {
+    id: 'course.assessment.submission.updateIndividualSuccess',
+    defaultMessage: 'Submission for {errors} updated successfully',
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/translations.ts:289]: you're right. At the end, I decided to employ the Chip that basically will switch between Saved and Unsaved to denote that a grade has been saved or not

File: client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx (hunk 10 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx:149]: it's not that I know it will be always false, but rather that it needs to be false.
this function will only be called when we want to update the grade for one question and not the entire submission, and hence we should never allow the experience points to be awarded as points, but rather as draft.

Only when the grader decided to publish the grade (by clicking on the button "Publish Grade" on the bottom of the page) then the points awarded will be awarded as points

Comment (ekowidianto) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx:149]: What if the submission is already in published mode and the exp is updated?

Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/index.jsx:149]: okay there are 2 possibilities this might happen, on how exp is updated:
1. exp is updated directly at the EXP Points box inside the submission. In this case, this function will contribute nothing to it, as it will never be invoked in the first place
2. it's updated because we modify the grade in one of the question. This one, to be honest, I was not aware that the grade can be updated even after it's been published. My apologies for that I'm unaware about this situation, and I have updated it accordingly. And you are correct that published might be True.

File: client/app/bundles/course/assessment/submission/containers/QuestionGrade.jsx (hunk 8 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/containers/QuestionGrade.jsx]: yeap.. already done, will push soon

File: client/app/bundles/course/assessment/submission/containers/QuestionGrade.tsx [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+      <Typography color="text.secondary" variant="body1">
+        {intl.formatMessage(translations.grade)}
+      </Typography>
+
+      {editable ? renderQuestionGradeField() : renderQuestionGrade()}
+    </Paper>
+  );
+};
+
+export default injectIntl(QuestionGrade);
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/containers/QuestionGrade.tsx:181]: Dont use injectIntl. use useTranslations

File: client/app/bundles/course/assessment/submission/containers/QuestionGrade.tsx [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+            <Chip
+              className="slot-1-neutral-400 border-slot-1 text-slot-1"
+              label={intl.formatMessage(translations.gradePrefilled)}
+              size="small"
+              variant="outlined"
+            />
+          </Tooltip>
+        )}
+
+        {dirty ? (
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/submission/containers/QuestionGrade.tsx:140]: Show that a grade is being saved as well

Comment (bivanalhar) [reply client/app/bundles/course/assessment/submission/containers/QuestionGrade.tsx:140]: I'm sorry but seems like I don't really understand your concern here. If the grade has been saved, it's indeed shown as "Saved", as follows
Screenshot 2023-10-03 at 11 06 46 PM

**Review by ekowidianto — COMMENTED** [review]
Some comments Thanks!

File: app/services/course/assessment/submission/update_service.rb [hunk 17]
```diff
@@ -153,15 +150,17 @@ def update_submission # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComple
           answer = @submission.answers.includes(:actable).find { |a| a.id == answer_params[:id].to_i }
           if answer && !update_answer(answer, answer_params)
             logger.error("Failed to update answer #{answer.errors.inspect}")
-            raise ActiveRecord::Rollback
+            render json: { errors: answer.errors }, status: :bad_request
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/services/course/assessment/submission/update_service.rb:153]: the main objective for rendering json over here instead of the previous place (in the update function) is basically because I want to give proper errors to the frontend (previously, the error occuring inside the answer won't be reflected properly because they're trying to get the error messages from the submission, which doesn't makes sense)

However, I just realised that Rollback plays the pivotal thing here to ensure that the transaction will be all aborted properly when it fails, so I am retrieving this raise rollback back..

File: client/app/api/course/Assessment/Submissions.js [hunk 18]
```diff
@@ -65,6 +65,13 @@ export default class SubmissionsAPI extends BaseAssessmentAPI {
     return this.client.get(`${this.#urlPrefix}/${submissionId}/edit`);
   }
 
+  updateGrade(submissionId, submissionFields) {
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/api/course/Assessment/Submissions.js:68]: the existing API handles properly the sending of the file data to the backend (using FormData), however it prevents the FE to send null value to the backend

the one I just created did the other way around; allowing FE to send null value to the backend, but failed to properly sending the file data to the backend

I didn't find any way to reconcile these 2, and that's why I created the separate API for general update compared with grade update

[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/api/course/Assessment/Submissions.js:68]: Then you should specify the field in updateGrade, not just pass the whole submissionFields. See code with `adaptedData`.

[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/api/course/Assessment/Submissions.js:68]: Then you can get the error out from the answer, and do something like `submission.errors.add(answer_error)` no?

File: app/services/course/assessment/submission/update_service.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (ekowidianto) [reply app/services/course/assessment/submission/update_service.rb:153]: Then you can get the error out from the answer, and do something like submission.errors.add(answer_error) no?

File: client/app/api/course/Assessment/Submissions.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/api/course/Assessment/Submissions.js]: got it.. I have restricted the fields in that updateGrade to contain only id and grade for each of the elements.
as for checking in the background, I have already did the check on this one, it's inside the file `app/models/course/assessment/answer.rb`, line 116 to 118

File: app/services/course/assessment/submission/update_service.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/services/course/assessment/submission/update_service.rb]: now you mentioned it, this is also another alternative that makes the code much cleaner.. I have updated it accordingly

File: app/models/course/assessment/answer.rb [hunk 19]
```diff
@@ -38,7 +38,8 @@ class Course::Assessment::Answer < ApplicationRecord
   validate :validate_grade, unless: :attempting?
   validates :submitted_at, presence: true, unless: :attempting?
   validates :submitted_at, :grade, :grader, :graded_at, absence: true, if: :attempting?
-  validates :grade, :grader, :graded_at, presence: true, if: :graded?
+  validates :grade, presence: { if: :graded?, message: 'Grade must not be empty if it\'s already graded' }
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/models/course/assessment/answer.rb]: done

**Review by ekowidianto — APPROVED** [review]
LGTM Thanks!
